src/monitor.h

```diff
@@ -59,6 +59,118 @@ typedef struct {
     long unsigned int cpu_total_time;
 } CpuUsage;
 
+class CpuUsageTracker {
+public:
+    /**
+     * Returns estimated CPU usage since last call to this function.
+     */
+    int GetCurrent(float* ucpu_usage, float* scpu_usage, long int * uticks, long int* sticks);
+    CpuUsageTracker();
+private:
+    int ReadCpuUsage(CpuUsage* result);
+    void CalculateCpuUsage(CpuUsage* cur_usage, CpuUsage* last_usage,
+                    float* ucpu_usage, float* scpu_usage, long int * uticks, long int* sticks);
+    CpuUsage lastUsage_;
+    CpuUsage currentUsage_;
+};
+
+typedef struct {
+    long unsigned int numCalls;
+    uint64_t cumulativeTime; //< in nanoseconds
+    uint64_t maxTime;        //< in nanoseconds
+} GCStat;
+
+/**
+ * Simple wrapper class to handle scoped lock for pre-existing libuv mutex
+ */
+class ScopedUVLock {
+ public:
+    ScopedUVLock(uv_mutex_t* m) : m_(m) { uv_mutex_lock(m_); }
+    ~ScopedUVLock() { uv_mutex_unlock( m_ ); }
+ private:
+    uv_mutex_t* m_;
+
+    ScopedUVLock();  //< can't create one of these without a reference to a mutex
+};
+
+/**
+ * encapsulates garbage collection stats per "interval" where
+ * interval is arbitrary time between calls to GetStats
+ **/
+class GCUsage {
+public:
+    GCUsage();
+    ~GCUsage();
+
+    /** Returns garbage collection (GC) stats since last call to this function.
+     *  Also resets the counters in order to start a new interval of GC stats
+     */
+    const GCStat EndInterval();
+
+    /**
+     * start and stop events get called from v8 gc prologue/epilogue
+     **/
+    void Start();
+    void Stop();
+
+ private:
+    GCStat     stats_;
+    uint64_t   startTime_;
+
+    // lock used to prevent reading/writing GCStat data structure at the same time
+    // by more than one thread
+    uv_mutex_t  lock_;
+
+    /* prevent copying/assigning */
+    GCUsage(const GCUsage&);
+    GCUsage& operator=(const GCUsage&);
+};
+
+
+class GCUsageTracker {
+public:
+    GCUsageTracker() {};
+    ~GCUsageTracker() {};
+
+    inline GCUsage* GetGCUsage( const v8::GCType type ) {
+        int collector = v8GCTypeToIndex( type );
+        assert( collector >= 0 && collector < kNumGCTypes );
+
+        return &usage_[ v8GCTypeToIndex( type )];
+    }
+
+    static const int kNumGCTypes = 2;  //< must match # of distinct v8::GCTypes in v8.h
```

Comment:
You probably looked into this, but is there a way to get this number from the v8 headers?